modules/dnn/src/model.cpp

```diff
@@ -637,6 +637,257 @@ void DetectionModel::detect(InputArray frame, CV_OUT std::vector<int>& classIds,
         CV_Error(Error::StsNotImplemented, "Unknown output layer type: \"" + lastLayer->type + "\"");
 }
 
+class YoloObjectDetector_Impl : public Model::Impl
+{
+    // Define the enum for YOLO versions
+    enum class YoloVersion {
+        YOLOVOC,
+        YOLOv3,
+        YOLOv4,
+        YOLOv5,
+        YOLOv6,
+        YOLOv7,
+        YOLOv8,
+        YOLOx,
+    };
+
+    Net    net;
+    Size   size;
+    Scalar mean;
+    Scalar scale = Scalar::all(1.0);
+    bool   swapRB = false;
+    bool   crop;
+    bool   nmsAcrossClasses;
+    float  padValue;
+    Mat    blob;
+    ImagePaddingMode paddingMode;
+    std::vector<String> outNames;
+    public:
+    YoloVersion yoloVersion;
+    bool   darknet;
+
+    public:
+    YoloObjectDetector_Impl() {
+        //nothing
+    }
+    YoloObjectDetector_Impl(const int version)
+    {
+        setYoloVersion(version);
+    }
+
+    /*virtual*/
+    void processFrame(InputArray frame, OutputArrayOfArrays outs)
+    {
+        CV_TRACE_FUNCTION();
+        if (size.empty())
+            CV_Error(Error::StsBadSize, "Input size not specified");
+
+        Image2BlobParams param;
+        param.scalefactor = scale;
+        param.size = size;
+        param.mean = mean;
+        param.swapRB = swapRB;
+        param.borderValue = padValue;
+        param.paddingmode = paddingMode;
+
+        Mat blob = dnn::blobFromImageWithParams(frame, param); // [1, 10, 10, 4]
+        net.setInput(blob);
+        net.forward(outs, outNames);
+    }
+
+    void setPaddingValue(const float padValue_){
+        padValue = padValue_;
+    }
+
+    void setPaddingMode(const ImagePaddingMode paddingmode_){
+        paddingMode = paddingmode_;
+    }
+    void setNmsAcrossClasses(bool value) {
+        nmsAcrossClasses = value;
+    }
+
+    void setNetworkType(const bool darknet_ = false){
+        darknet = darknet_;
+    }
+
+    void setYoloVersion(const int versionInt){
+       if (
+            versionInt == static_cast<int>(YoloVersion::YOLOVOC) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv3) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv4) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv5) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv6) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv7) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOv8) ||
+            versionInt == static_cast<int>(YoloVersion::YOLOx)
+            ) {
+            yoloVersion = static_cast<YoloVersion>(versionInt);
```

Comment:
You should assign explicit int values to the enum. Otherwise `setYoloVersion(5)` means not `YOLOv5` but `YOLOv7`

Reply:
Done. But now yolox is index as 9. Not sure if this is a nice solution

Reply:
Can we skip explicit version setup for now but detect `YOLOv8` (as far as I can see there is only one version of YOLO which differs from others in `yoloPostProccess`? For example, by the shapes of `detections[0]`?

Reply:
yes, for yolov8 and darknet models as well. For darknet models there is explicit flag